### ISSUE TITLE
ACP: Reverse internals (`copy_from_reversed_slice` & `swap_with_reversed_slice`)

### ISSUE THREAD

**Issue by clarfonthey** [rust-lang/libs-team#68]
# Proposal

## Problem statement

`slice::reverse` is a fundamental operation on slices that has been optimized to ensure that it runs as fast as possible. However, because of its nature, this operation only works on contiguous slices, and if someone were to write a `reverse` implementation for their own data structure which is not contiguous, they would have to essentially rewrite the implementation of `slice::reverse` for themself, with changes to account for their own data structure's layout. The goal would be to expose more methods that could help avoid rewriting the code that's already used in libstd.

## Motivation, use-cases

A large number of use cases revolve around writing a slice in reversed order, or swapping two slices element-by-element in reversed order. If we call these operations `copy_reversed` and `swap_reversed` for the sake of argument, then a bunch of different operations can be written in terms of them. For example:

Reversing a slice requires splitting the slice into halves (ignoring the odd center element if it exists), then running `swap_reversed` on the two halves.

Reversing a ring buffer is similar to a regular slice reversal, except you have to split the buffer into three parts: one part from each half of the ring buffer to be `swap_reversed`, then a leftover bit to be reversed normally.

If data is split into several buffers, the reversal is done very similarly to the ring buffer case, except potentially several times.

In my case, I was interested in a `copy_reversed` intrinsic to convert between endianness of bigint formats, since I wanted to copy a big-endian buffer into a new little-endian one, but the current methods on `slice` require me to either rewrite my own `reverse` implementation to copy instead, or to perform the copy and reverse separately, which performs twice as many writes even after LLVM optimization.

## Solution sketches

I basically want to propose adding the below three functions. I've included "definitions" involving iterators because they feel the most understandable, but the actual implementations will borrow code from libstd.

```rust
use core::mem;

impl<T> [T] {
    pub fn copy_from_reversed_slice(&mut self, other: &[T]) where T: Copy {
        assert!(self.len() == other.len());
        self.iter_mut().zip(other.iter().rev()).for_each(|dest, src| *dest = *src);
    }
    pub fn clone_from_reversed_slice(&mut self, other: &[T]) where T: Clone {
        assert!(self.len() == other.len());
        self.iter_mut().zip(other.iter().rev()).for_each(|dest, src| *dest = src.clone());
    }
    pub fn swap_with_reversed_slice(&mut self, other: &mut [T]) {
        assert!(self.len() == other.len());
        self.iter_mut().zip(other.iter_mut().rev()).for_each(|dest, src| mem::swap(dest, src));
    }
}
```

We could potentially extend this to include methods in the `ptr` module too, like:

```rust
pub unsafe fn swap_nonoverlapping_reversed<T>(x: *mut T, y: *mut T, count: usize);
pub unsafe fn copy_nonoverlapping_reversed<T>(src: *const T, dest: *mut T, count: usize);
```

However, I'm not sure whether this would be desired. Really, these two methods perform identical things in the end, with the exception that the `ptr` methods will never panic due to a mismatched length, because this requirement is offloaded as a safety contract instead.

### Similar implementations

C++2017 added a `reverse_copy` function: https://en.cppreference.com/w/cpp/algorithm/reverse_copy

I couldn't find any other references to a similar API, but I'm not sure many exist.

### Benchmark

As part of writing this, I performed a benchmark between four possible implementations of `copy_reverse`, namely:

1. Simple `dest.copy_from_slice(src)`, then `dest.reverse()`
2. `src.reverse()`, then `dest.copy_from_slice(src)`
3. For `i in 0..len`, `dest[i] = src[n - 1 - i]`
4. For `i in 0..len`, `dest[n - 1 - i] = src[i]`

As expected, the difference between 3 and 4 was statistically insignificant. Also as expected, this is the fastest approach, and also the one that the existing `reverse` uses in libstd.

Interestingly, 1 performs faster than 2 in almost all cases. I didn't take much of a look at the generated IR, but my guess is that LLVM could detect the redundant writes and remove them in the first case, but not the second.

If anyone's actually interested, I can share the code somewhere. But the bottom line is that there is a noticeable difference between the implementations, and it's not immediately obvious to a potential implementer which one is the fastest. Having a singular operation instead of composing operations makes the correct choice more clear.

## What happens now?

This issue is part of the libs-api team [API change proposal process]. Once this issue is filed the libs-api team will review open proposals in its weekly meeting. You should receive feedback within a week or two.

[API change proposal process]: https://std-dev-guide.rust-lang.org/feature-lifecycle/api-change-proposals.html

**Comment by scottmcm** [comment]
I assume there should be a `.rev()` in one side of the zip for `swap_with_reversed_slice`?

You mentioned this in passing, but I'll emphasize that `[T]::reverse` works today by splitting it into two slices and reverse-swapping those, which suggests it might be a more primitive operation:

https://github.com/rust-lang/rust/blob/75dbd5b8c3e3e3e7673caa890422e654d39b815e/library/core/src/slice/mod.rs#L702-L718

**Comment by clarfonthey** [comment]
Yes, I don't know how I managed to post this the first time without the `rev()` in the implementations.

I mostly proposed the copying since while it's not fundamental to swapping slices, it could be used for other algorithms and will have a rather similar feel. However, I understand that it may be less reasonable to include, which is also why I figured that going through the ACP process and discussing it was a good idea.
